### PR TITLE
[11.0][ADD] Auth SAML create users

### PR DESCRIPTION
**Pull request by eilst**

### CODE REVIEW

**Comment by eilst** [comment]
I could add a docker-compose.yml with an idp configured with odoo as sp for functional tests. Is it a good idea?

**Comment by eilst** [comment]
Hey @max3903 @pedrobaeza @vincent-hatakeyama, I added this feature of giving the option to automatically create a user when he logs in with an idp. I base this pr on @vincent-hatakeyama, which fixes some important issues that exist on 11.0.

**Comment by eilst** [comment]
The runbot fails with when building the docker image 
ERROR: /bin/sh: 1: mysql_config: not found
    Traceback (most recent call last):...
      File "/tmp/pip-install-lro6f39d/mysqlclient/setup_posix.py", line 29, in mysql_config
        raise EnvironmentError("%s not found" % (_mysql_config_path,))
    OSError: mysql_config not found
I think is something with the dockerfile, right?

**Review by lgharib — APPROVED** [review]

File: oca_dependencies.txt [hunk 1]
```diff
@@ -1,2 +1 @@
-# base_user_role
-server-backend
+
```
Comment (eilst) [inline oca_dependencies.txt]:
```suggestion
server-backend
```

File: auth_saml_create_user/models/auth_saml.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Copyright (C) 2010-2016 XCG Consulting <http://odoo.consulting>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
```
Comment (rimbendhaou) [inline auth_saml_create_user/models/auth_saml.py:2]: why this file has different copyright and license

File: auth_saml_create_user/models/auth_saml.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+from odoo import fields, models
+
+
+class AuthSamlProvider(models.Model):
+    _inherit = ['auth.saml.provider']
```
Comment (rimbendhaou) [inline auth_saml_create_user/models/auth_saml.py]:
```suggestion
    _inherit = 'auth.saml.provider'
```

File: auth_saml_create_user/models/res_users.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        return self.env['auth.saml.provider'].browse(provider).create_user
+
+    def create_user(self, saml_uid, provider):
+        _logger.debug("Creating new Odoo user \"%s\" from SAML" % saml_uid)
+        SudoUser = self.env['res.users'].sudo()
+        new_user = SudoUser.create({
+            'name': saml_uid,
+            'login': saml_uid,
+            'saml_provider_id': provider,
+            'company_id': self.env['res.company'].sudo().browse(1).id,
```
Comment (rimbendhaou) [inline auth_saml_create_user/models/res_users.py:37]: browse(1)? better to not use id directly, ids could be changed

Comment (eilst) [reply auth_saml_create_user/models/res_users.py:37]: In fact, the feature that would extend and improve this module is to allow the creation of 'template users'.Right now, the "My Company will be used". So I will create an issue for this as a feature request.

File: auth_saml_create_user/models/auth_saml.py (hunk 2 above)
Comment (eilst) [inline auth_saml_create_user/models/auth_saml.py:2]:
```suggestion
# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
```

Comment (eilst) [reply auth_saml_create_user/models/auth_saml.py:2]: Done, thanks

**Review by wbeverly — APPROVED** [review]
Approved!

**Comment by eilst** [comment]
@bodedra @max3903 This module is ready with unit tests added. Could you merge it?

File: auth_saml_create_user/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# © 2019 Savoir-faire Linux
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+{
+    'name': "Auth SAML Create User",
+
```
Comment (bodedra) [inline auth_saml_create_user/__manifest__.py]: We can remove empty line.

File: auth_saml_create_user/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    'summary': """
+        This module extends the functionality of Auth SAML to support
+        the automatic creation of SAML users when they don't exist in odoo.""",
+    'author': 'Savoir-faire Linux, Odoo Community Association (OCA)',
+    'website': 'https://github.com/OCA/server-auth',
+    'license': 'LGPL-3',
+    'category': 'Tools',
+    'version': '11.0.1.0.0',
+    'depends': [
+        'base',
```
Comment (bodedra) [inline auth_saml_create_user/__manifest__.py]: Do we need `base` dependency ?

File: auth_saml_create_user/models/res_users.py (hunk 4 above)
Comment (bodedra) [reply auth_saml_create_user/models/res_users.py:37]:
```suggestion
            'company_id': self.env.ref('base.main_company').id
```

**Review by bodedra — COMMENTED** [review]
Add manor comments.

File: auth_saml_create_user/__manifest__.py (hunk 5 above)
Comment (eilst) [reply auth_saml_create_user/__manifest__.py]: Done

File: auth_saml_create_user/__manifest__.py (hunk 6 above)
Comment (eilst) [reply auth_saml_create_user/__manifest__.py]: Right, done thanks

**Comment by vincent-hatakeyama** [comment]
Why is there a mix of LGPL and AGPL code in the module? It has always been AGPL, I don’t think LGPL can be added into it.

I also don’t think you can change license of the code like indicated in https://github.com/OCA/server-auth/pull/113#discussion_r306542699

**Comment by eilst** [comment]
> Why is there a mix of LGPL and AGPL code in the module? It has always been AGPL, I don’t think LGPL can be added into it.
> 
> I also don’t think you can change license of the code like indicated in [#113 (comment)](https://github.com/OCA/server-auth/pull/113#discussion_r306542699)

Right, I rebased the branch with the correct license. In other projects that I work they use LGPL, so I did a mistake in some file headers. 
I also read this in https://odoo-community.org/page/faq :
Why most of the OCA code base is AGPL ?
Because the AGPL license is the only license considering someone connecting to Odoo as a user. It protects companies from being locked in by their Odoo hosting provider. Finally, Odoo was fully AGPL for years and lots of devs has been released under the license by then.

**Comment by eilst** [comment]
Now, everything should be fine. License, tests, code coverage, travis checks, corrections from comments, main feature. Could you merge it please @pedrobaeza?

**Comment by moylop260** [comment]
/ocabot merge

Retrying... since that the error is about an unittest retrieving a URL

Let's to trying again

**Comment by moylop260** [comment]
/ocabot merge

There is still a flaky error

**Comment by moylop260** [comment]
@eilst 
I just detected the red error of the CI.

Fixing in the following PR:
 - https://github.com/OCA/server-auth/pull/149

**Pull request merged**